pil-analyzer/src/traits_resolver.rs

```diff
@@ -1,131 +1,85 @@
 use powdr_ast::{
-    analyzed::{
-        Expression, FunctionValueDefinition, Identity, PolynomialReference, Reference, Symbol,
-        TypedExpression,
-    },
+    analyzed::{Expression, PolynomialReference},
     parsed::{
         types::{TupleType, Type},
-        visitor::AllChildren,
-        SelectedExpressions, TraitImplementation,
+        TraitImplementation,
     },
 };
 use std::{collections::HashMap, sync::Arc};
 
-type SolvedImpl = ((String, Vec<Type>), Arc<Expression>);
+use crate::type_unifier::Unifier;
 
-pub struct TraitsResolver<'a> {
-    definitions: &'a HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
-    identities: &'a Vec<Identity<SelectedExpressions<Expression>>>,
+/// Resolves trait implementations for all definitions and identities.
+/// Checks references for calls to trait implementations
+/// and unifies them using their type args.
+///
+/// Returns a nested map of trait name to a map of type args to the expression.
+pub fn traits_resolution(
+    references: Vec<PolynomialReference>,
+    trait_impls: &HashMap<String, Vec<TraitImplementation<Expression>>>,
+) -> HashMap<String, HashMap<Vec<Type>, Arc<Expression>>> {
+    TraitsResolver::new().resolve_traits(references, trait_impls)
 }
 
-impl<'a> TraitsResolver<'a> {
-    pub fn new(
-        definitions: &'a HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
-        identities: &'a Vec<Identity<SelectedExpressions<Expression>>>,
-    ) -> Self {
-        Self {
-            definitions,
-            identities,
-        }
+type SolvedImpl = ((String, Vec<Type>), Arc<Expression>);
+
+struct TraitsResolver {}
+
+impl TraitsResolver {
+    pub fn new() -> Self {
+        Self {}
     }
 
     pub fn resolve_traits(
-        &'a self,
+        &self,
+        references: Vec<PolynomialReference>,
```

Comment:
As far as I understand it, this function can fail. Not in the case where the reference cannot be uniquely resolved (this should be handled generically by checking that each pair of impls does not unify and is done before this step), but in the case where there is no impl for the type args. Because of that, I think we should have source references here and return a vector of errors (there is no reason to fail on the first error).

Reply:
But this can be a new PR.

Reply:
The problem is that there is not always an implementation for every reference. Since we take every reference without distinction, they must `fail` in some silent way.

Reply:
Sure, and I think this is the main place where this can be detected, so it should not fail silently, but maybe I'm wrong.

Reply:
The problem I see is that when we receive a reference that has no implementation, let's say any variable `F::x`, it will never find an implementation and will always fail.